File: backend/src/main/java/com/example/kolla/services/DepartmentService.java
package com.example.kolla.services;

import com.example.kolla.dto.CreateDepartmentRequest;
import com.example.kolla.dto.UpdateDepartmentRequest;
import com.example.kolla.responses.DepartmentResponse;

import java.util.List;

/**
 * Department management service interface.
 * Requirements: 12.1–12.8
 */
public interface DepartmentService {

    List<DepartmentResponse> listDepartments();

    DepartmentResponse getDepartmentById(Long id);

    DepartmentResponse createDepartment(CreateDepartmentRequest request);

    DepartmentResponse updateDepartment(Long id, UpdateDepartmentRequest request);

    void deleteDepartment(Long id);
}

File: frontend/src/services/departmentService.ts
/**
 * departmentService.ts — wraps all department-related API calls.
 * Requirements: 12.1–12.8
 */

import api from './api'
import type { ApiResponse } from '../types/api'

export interface DepartmentDto {
  id: number
  name: string
  description?: string
  createdAt?: string
  updatedAt?: string
}

export interface CreateDepartmentRequest {
  name: string
  description?: string
}

export interface UpdateDepartmentRequest {
  name?: string
  description?: string
}

export async function listDepartments(): Promise<ApiResponse<DepartmentDto[]>> {
  const res = await api.get<ApiResponse<DepartmentDto[]>>('/departments')
  return res.data
}

export async function getDepartment(id: number): Promise<ApiResponse<DepartmentDto>> {
  const res = await api.get<ApiResponse<DepartmentDto>>(`/departments/${id}`)
  return res.data
}

export async function createDepartment(
  data: CreateDepartmentRequest,
): Promise<ApiResponse<DepartmentDto>> {
  const res = await api.post<ApiResponse<DepartmentDto>>('/departments', data)
  return res.data
}

export async function updateDepartment(
  id: number,
  data: UpdateDepartmentRequest,
): Promise<ApiResponse<DepartmentDto>> {
  const res = await api.put<ApiResponse<DepartmentDto>>(`/departments/${id}`, data)
  return res.data
}

export async function deleteDepartment(id: number): Promise<void> {
  await api.delete(`/departments/${id}`)
}
